library/default-effects-extra.scd

```diff
@@ -1,390 +1,450 @@
+	signal = signal + (octersub * oct2 * sub) + (octersubsub * oct3 * sub);
+	ReplaceOut.ar(out, signal);
+}, [\ir, \ir, \ir, \ir]).add;
+
+// Ring modulation with `ring` (modulation amount), `ringf` (modulation frequency), and `ringdf` (slide
+// in modulation frequency)
+~dirt.addModule('ring', { |dirtEvent|
+	dirtEvent.sendSynth("dirt_ring" ++ ~dirt.numChannels,
+		[
+			ring: ~ring,
+			ringf: ~ringf,
+			ringdf: ~ringdf,
+			out: ~out
+		]
+	)
+}, { ~ring.notNil });
+
+SynthDef("dirt_ring" ++ ~dirt.numChannels, { |out, ring = 0, ringf = 0, ringdf|
+	var signal, mod;
+	signal = In.ar(out, ~dirt.numChannels);
+	mod = ring * SinOsc.ar(Clip.kr(XLine.kr(ringf, ringf + ringdf), 20, 20000));
+	signal = ring1(signal, mod);
+	ReplaceOut.ar(out, signal);
+}, [\ir, \ir, \ir, \ir]).add;
+
+// A crunchy distortion with a lot of high harmonics, the only parameter is `distort`
+~dirt.addModule('distort', { |dirtEvent|
+	dirtEvent.sendSynth("dirt_distort" ++ ~dirt.numChannels,
+		[
+			distort: ~distort,
+			out: ~out
+		]
+	)
+}, { ~distort.notNil });
+
+SynthDef("dirt_distort" ++ ~dirt.numChannels, { |out, distort = 0|
+	var signal, mod;
+	signal = In.ar(out, ~dirt.numChannels);
+	mod = CrossoverDistortion.ar(signal, amp: 0.2, smooth: 0.01);
+	mod = mod + (0.1 * distort * DynKlank.ar(`[[60,61,240,3000 + SinOsc.ar(62,mul: 100)],nil,[0.1, 0.1, 0.05, 0.01]], signal));
+	mod = (mod.cubed * 8).softclip * 0.5;
+	mod = SelectX.ar(distort, [signal, mod]);
+	ReplaceOut.ar(out, mod);
+}, [\ir, \ir]).add;
+
+// Spectral delay
+~dirt.addModule('spectral-delay', { |dirtEvent|
+	dirtEvent.sendSynth('spectral-delay' ++ ~dirt.numChannels,
+		// OPTIONAL
+		// passing this array of parameters could be left out,
+		// but it makes it clear what happens and it is also more
+		// effecient to explicitly specify the arguments
+		[
+			xsdelay: ~xsdelay,
+			tsdelay: ~tsdelay,
+			out: ~out
+		]
+	)
+}, { ~tsdelay.notNil or: { ~xsdelay.notNil }});
+
+SynthDef("spectral-delay" ++ ~dirt.numChannels, { |out, tsdelay = 0.5, xsdelay = 0.5|
+
+	var signal, delayTime, delays, freqs, filtered;
+	var size = 16;
+	var maxDelayTime = 0.2;
+	signal = In.ar(out, ~dirt.numChannels);
+	delayTime = tsdelay * maxDelayTime;
+	filtered = (1..size).sum { |i|
+		var filterFreq = i.linexp(1, size, 40, 17000);
+		var sig = BPF.ar(signal, filterFreq, 0.005);
+		// the delay pattern is determined from xsdelay by bitwise-and:
+		DelayN.ar(sig, maxDelayTime, i & xsdelay * (1 / size) * delayTime )
+	};
+	signal = signal * 0.2 + (filtered * 4); // this controls wet / dry
+	ReplaceOut.ar(out, signal)
+}, [\ir, \ir, \ir]).add;
+
+// Spectral freeze
+~dirt.addModule('spectral-freeze', { |dirtEvent|
+	dirtEvent.sendSynth('spectral-freeze' ++ ~dirt.numChannels,
+		[
+			freeze: ~freeze,
+			out: ~out
+		]
+	)
+}, { ~freeze.notNil } );
+
+SynthDef("spectral-freeze" ++ ~dirt.numChannels, { |out, freeze|
+	var signal, chain, in;
+
+	signal = In.ar(out, ~dirt.numChannels);
+
+	if (~dirt.numChannels > 1 , {			
```

Comment:
I don't see why these case distinctions are necessary here at all - what was it that didn't work?

Reply:
if (~dirt.numChannels > 1 , {			
		chain = Array.fill(signal.size, { |i| FFT(LocalBuf(2048), signal[i])});
	} , {
		chain = FFT(LocalBuf(2048), signal);
	});

isn't it that when the dimension of signal is 1 (a vector) signal[0] is a sample ?

Reply:
No, this is not how it works. SuperCollider operates on Objects that represent abstract signal processors, you never have access to a sample directly. Much like Haskell actually. In our case here, `signal` is a `UGen` object. A single object has size 1, so it will correctly behave like an array with one element.